duckduckgo/macos-browser
# Duck Player (native)

Task/Issue URL: https://app.asana.com/0/1179596784783755/1202854196822746/f

**Description**:
This change allows users to watch YouTube videos in a private player loaded from a bundled HTML file,
embedding "private" YouTube iframe (loading videos from `www.youtube-nocookie.com`) that doesn't
store cookies (and doesn't display personalized ads). Duck Player can be turned off completely,
enabled on demand (accessible via video overlay buttons) or enabled automatically for every YouTube video.

Since Duck Player adds significant complexity to (already complex) Tab navigation, all the logic
has been abstracted into a helper class called `Private Player`. The points of contact with current code are:
* TabContent value calculation.
* Tab URL loading,
* WebView back navigation,
* Privacy feed items.

Duck Player is loaded using a custom scheme handler and works with URLs like `duck://player/VIDEOID?t=TIME`.
On macOS 12 and above it uses WKWebView's `loadSimulatedRequest` API to inject `Referer` HTTP header
into the private player request, to support playing videos containing copyrighted content (e.g. music videos).
This is not available on macOS 11, so some videos won't play in Duck Player.

The whole feature is temporarily disabled on macOS Catalina, while we're figuring out issues around loading
YouTube video overlays, and security-related problems (it's done using `PrivatePlayer.isAvailable` – can later
be used to disable the feature completely using e.g. remote config).



**Steps to test this PR**:
Please not that JS and HTML is being reviewed in separate pull requests and this PR is only about the native code 🙏 

"Ask every time" mode:
1. Go to youtube.com.
1. Notice that video thumbnails have a Dax icon overlayed in the top-left corner.
1. Open a video by clicking on it (not on the Dax icon).
1. Verify that the video overlay is presented asking you to try Duck Player.
1. Select "Watch in Duck Player" and verify that the video opens in Duck Player.
1. Verify that Settings button and "Watch in YouTube" button work as expected.
1. Verify that clicking on video title opens YouTube video (without Duck Player) in a new tab. 
1. Pause the video in Duck Player and click on some recommendation; verify that it opens in the same Duck Player.

Settings:
1. Have Duck Player and browser Settings opened in 2 tabs.
1. Notice that if you're in "Show option to use Duck Player on hover" mode, the Duck Player displays "Always open YouTube videos in Duck Player" checkbox.
1. Click the checkbox, verify that it's gone and that "Always open YouTube videos in Duck Player" mode is selected in Settings
1. Switch back to "Show option ...", verify that the checkbox is visible again in Duck Player and it's unchecked.

"Always open" mode:
1. Enter "always open" mode, e.g. via Settings
1. Go to youtube.com
1. Verify that videos don't show overlays
1. Click any video and verify that you're being redirected to Duck Player (some visual glitches are expected)
1. While in Duck Player, click back button and verify that you're successfully redirected to youtube.com
1. While in Duck Player, swipe to navigate back and verify that you're successfully redirected to youtube.com (visual glitches expected)
1. While in Duck Player, click "Watch in YouTube" and verify that the video opens in youtube.com.

"Disabled" mode:
1. Disable Duck Player in settings
1. Verify that youtube.com doesn't show overlays
1. Verify that videos are played on youtube.com

Other:
1. Open a video in Duck Player
1. Quit the app and reopen
1. Verify that it reopens Duck Player in both "always ask" and "always open in Duck Player" modes.
1. Verify that Privacy Feed on home page displays "Duck Player" as "domain" and it's not clickable, and for individual videos their titles are displayed.

**Testing checklist**:


---
###### Internal references:
[Software Engineering Expectations](https://app.asana.com/0/59792373528535/199064865822552)
[Technical Design Template](https://app.asana.com/0/59792373528535/184709971311943)
**When ready for review, remember to post the PR in MM**
